cmd/tiller-proxy/internal/handler/apprepos_handler.go

```diff
@@ -143,23 +173,98 @@ func (a *appRepositoriesHandler) Create(w http.ResponseWriter, req *http.Request
 	}
 
+	secret := secretForRequest(appRepoRequest, *appRepo)
+	if secret != nil {
+		_, err = clientset.CoreV1().Secrets(a.kubeappsNamespace).Create(secret)
+		if err != nil {
+			if statusErr, ok := err.(*errors.StatusError); ok {
+				status := statusErr.ErrStatus
+				log.Infof("unable to create app repo secret: %v", status.Reason)
+				http.Error(w, status.Message, int(status.Code))
+			} else {
+				log.Errorf("unable to create app repo secret: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return
+		}
+	}
+
 	w.WriteHeader(http.StatusCreated)
 	w.Write([]byte("OK"))
 }
 
-// appRepositoryForRequestData takes care of parsing the request data into an AppRepository.
-func appRepositoryForRequestData(body io.ReadCloser) (*v1alpha1.AppRepository, error) {
-	var appRepoRequest appRepositoryRequest
-	err := json.NewDecoder(body).Decode(&appRepoRequest)
-	if err != nil {
-		return nil, err
+// appRepositoryForRequest takes care of parsing the request data into an AppRepository.
+func appRepositoryForRequest(appRepoRequest appRepositoryRequest) *v1alpha1.AppRepository {
+	appRepo := appRepoRequest.AppRepository
+
+	var auth v1alpha1.AppRepositoryAuth
+	if appRepo.AuthHeader != "" || appRepo.CustomCA != "" {
+		secretName := fmt.Sprintf("apprepo-%s-secrets", appRepo.Name)
+		if appRepo.AuthHeader != "" {
+			auth.Header = &v1alpha1.AppRepositoryAuthHeader{
+				SecretKeyRef: corev1.SecretKeySelector{
+					Key: "authorizationHeader",
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: secretName,
+					},
+				},
+			}
+		}
+		if appRepo.CustomCA != "" {
+			auth.CustomCA = &v1alpha1.AppRepositoryCustomCA{
+				SecretKeyRef: corev1.SecretKeySelector{
+					Key: "ca.crt",
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: secretName,
+					},
+				},
+			}
+		}
 	}
+
 	return &v1alpha1.AppRepository{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: appRepoRequest.AppRepository.Name,
+			Name: appRepo.Name,
 		},
 		Spec: v1alpha1.AppRepositorySpec{
-			URL: appRepoRequest.AppRepository.URL,
+			URL:                appRepo.RepoURL,
+			Type:               "helm",
+			Auth:               auth,
+			SyncJobPodTemplate: appRepo.SyncJobPodTemplate,
+			ResyncRequests:     appRepo.ResyncRequests,
 		},
-	}, nil
+	}
+}
+
+// secretForRequest takes care of parsing the request data into a secret for an AppRepository.
+func secretForRequest(appRepoRequest appRepositoryRequest, appRepo v1alpha1.AppRepository) *corev1.Secret {
+	appRepoDetails := appRepoRequest.AppRepository
+	secrets := map[string]string{}
+	if appRepoDetails.AuthHeader != "" {
+		secrets["authorizationHeader"] = appRepoDetails.AuthHeader
+	}
+	if appRepoDetails.CustomCA != "" {
+		secrets["ca.crt"] = appRepoDetails.CustomCA
+	}
+
+	if len(secrets) == 0 {
+		return nil
+	}
+	blockOwnerDeletion := true
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("apprepo-%s-secrets", appRepoDetails.Name),
```

Comment:
this secret name is generated twice, better to do it in a function so we avoid miss configurations in the future?

Reply:
Also updated. Thanks.